### ISSUE TITLE
Copy files from mvn machine to container running on 'remote' docker host

### ISSUE THREAD

**Issue by janssk1** [fabric8io/docker-maven-plugin#758]
I'm running mvn (plugin version 0.20.1) and docker (latest) on separate machines. 
I would like to push a 'test artifact' (in form of a maven dependency) to my production container during IT tests, so that it becomes locally available in the container during tests

I see a few options:
* Mount the test artifact as a volume in the production container. Since docker runs on another machine, i did not manage to get this working
* Create a dedicated image + container for the test artifact and then mount that container as a volume in the production container. Have not tried it yet. I guess it should work, but it feels quite heavy. 
* Rely on docker cp support to push the artifact in the production container once its started. Although raw docker supports this, I have not found a way to do this using the plugin 

Is the dedicated image+container the only working option or am i missing something ? 

Thanks, 
Koen

**Comment by rhuss** [comment]
You are right, the is currently no way to use a similar functionality like `docker cp` although this might be good addition. 

For 'deploying' artefacts we currently two approaches:

* Linking a data container containing the artefact into e.g. an application server container which then gets deployed
* Use the application server as a base image and add the arterfact into the deployment directory, build the image, then start the combined image for the test.

Both are not so involved, you can check the samples/data-jolokia-demo for both approaches (in different profiles).

Does this answer your question ?

**Comment by janssk1** [comment]
Ok.  When creating a data container, I hit another issue. My data container does not have any cmd (based on scratch), so it can not be started. 
However, the plugin always starts a container. AFAIK, there is no way to create but not start a container. that would be very useful for 'volume containers'.

**Comment by rhuss** [comment]
Agreed that this would make sense. However, I still would have to think about how to add this that it fits into the UI. Any ideas? (I don't want to introduce a top-level goal for this particular functionality, though). Maybe an option in the `<run>` configuration could do it ...

**Comment by rgbj** [comment]
First things first, thanks for your work on that very useful plugin. I'm chiming in because I, too, would need some kind of `docker cp` functionality: I use docker to run some integration tests on my artifact. I need to get the tests results back from the container after the tests run. For now, a volume mount (e.g. `<volume>${project.build.directory}:/somewhere</volume>`) allows me to get those tests results, _when the docker host is my maven build host_. But this setup won't allow me to use another, separate docker host.

As to the UI: I hear you say you don't want another top-level goal and I can't guess why not, but I don't see how to avoid it. Here is the way I map the plugin execution in the maven lifecycle:

```
                       <executions>
                            <execution>
                                <id>build</id>
                                <phase>package</phase>
                                <goals>
                                    <goal>build</goal>
                                </goals>
                            </execution>
                            <execution>
                                <id>it-start</id>
                                <phase>pre-integration-test</phase>
                                <goals>
                                    <goal>start</goal>
                                </goals>
                            </execution>
                            <execution>
                                <id>it-stop</id>
                                <phase>post-integration-test</phase>
                                <goals>
                                    <goal>stop</goal>
                                </goals>
                            </execution>
                        </executions>
```

For my use-case, I'd need to be able to specify when (as well as what, obviously) to `docker cp`. If it's not a top-level goal, then the configuration needs to allow to say _when_ to run that task, and that's basically the maven lifecycle's job.. It would seem natural to me to write something like:

```
                       <executions>
                            <execution>
                                <id>build</id>
                                <phase>package</phase>
                                <goals>
                                    <goal>build</goal>
                                </goals>
                            </execution>
                            <execution>
                                <id>it-start</id>
                                <phase>pre-integration-test</phase>
                                <goals>
                                    <goal>start</goal>
                                </goals>
                            </execution>
                            <execution>
                                <id>it-stop</id>
                                <phase>post-integration-test</phase>
                                <goals>
                                    <goal>copy</goal>
                                    <goal>stop</goal>
                                </goals>
                            </execution>
                        </executions>
```

with some part of the image `<run>` configuration specifying _what_ to copy, I guess with some sort of assembly or assembly-like syntax.

Anyway, thanks again for your work!

**Comment by rhuss** [comment]
@rgbj I see your point and it might make sense to introduce a `copy` coal for the use case to extract files from the integration test containers. 

With

> (I don't want to introduce a top-level goal for this particular functionality, though)

I meant to not introduce a goal for creating but not starting a container as described in https://github.com/fabric8io/docker-maven-plugin/issues/758#issuecomment-293837071 (which btw would be _also_ something quite usable ;-)

**Comment by mabrarov** [comment]
What about https://github.com/fabric8io/docker-maven-plugin/issues/752#issuecomment-716097260? Will that work for you case? 

I believe it should work if used like this:

```xml
<executions>
    <execution>
        <id>build</id>
        <phase>package</phase>
        <goals>
            <goal>build</goal>
        </goals>
    </execution>
    <execution>
        <id>it</id>
        <phase>integration-test</phase>
        <goals>
            <goal>start</goal> <!-- should define http://dmp.fabric8.io/#start-wait like https://github.com/mabrarov/maven-docker-builder/blob/feature/docker-maven-plugin_app_tar/app/pom.xml#L48 -->
            <goal>copy</goal>  <!-- performs copying just for containers started with start goal -->
            <goal>stop</goal>  <!-- stops only containers started with start goal -->
        </goals>
    </execution>
</executions>
```

**Comment by rhuss** [comment]
Unfortunately I'm in kind of stand-by maintenance for this plugin, since my focus shifted considerably. I'm still help to integrate new features (until a new maintainer is found), but I won't work on new features like a `copy` goal. As you can see from the lag of my responses, I also only work sporadically (but then in bulk) on PR and open issues.

Sorry for that, but as mentioned if you are fancy to contribute I'm happy to help to integrate.
